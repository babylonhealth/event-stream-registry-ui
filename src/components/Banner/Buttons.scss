@import "../EventStreamRegistry/_variables.scss";
@import 'material-icons/iconfont/material-icons.scss';

.buttons {
    position: absolute;
    float: right;
    right: 10px;
    top: 0;
}

.buttons span {
    vertical-align: middle;
    padding-left: 3px;
    margin: 0;
    text-align: center;
    font-size: 15px;
    color: $lightGrey;
    cursor: pointer;
    z-index: 100;
}

.buttons span:hover {
    color: $darkGrey;
}

.buttons span.clicked {
    color: $darkGrey;
}
