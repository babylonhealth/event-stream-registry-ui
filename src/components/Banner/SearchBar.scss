@import "../EventStreamRegistry/_variables.scss";

input {
  outline: 0;
  font-family: "Courier New", sans-serif;
  font-size: 12px;
  border-color: $lightGrey;
  border-width: 0 0 1px;
  padding-bottom: 4px;
  padding-left: $lpad;
  height: 12px;
  width: 100%
}
