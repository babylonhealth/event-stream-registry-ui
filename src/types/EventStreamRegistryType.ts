/**
 * The Event Stream Registry dataset.
 *
 * @param node - contains the set of services and topics in the Event Stream Registry.
 * @param links - defines the connectedness of the nodes.
 * @type {EventStreamRegistryType}
 */
export interface EventStreamRegistryType {
    nodes: EventStreamRegistryNode[],
    links: EventStreamRegistryLink[]
}

/**
 * A single Event Stream Registry node - either a topic or a service.
 *
 * @param id - the node id
 * @param type - the node type, either "SERVICE" or "TOPIC"
 * @param metadata - a map node metadata, the key is the metadata name, the value is a Metadata object
 * @param urls - a list of associated node urls
 * @type {EventStreamRegistryNode}
 * */
export interface EventStreamRegistryNode {
    id: string,
    type: NodeType,
    metadata: Record<string, Metadata>,
    urls: Url[],
}

/**
 * A metadata value.
 *
 * @param value - the metadata value
 * @param type - either 'NUMBER' or 'STRING', the metadata type, the type defines allowed filtering operations
 * @param url? - an associated url, renders values as links
 * @param status? - either 'GOOD' or 'BAD', indicates the "health" of a metric, e.g. a high consumer lag might be
 * consider 'unhealthy` of `BAD`, renders values `green` and `red`.
 * @type {Metadata}
 * */
export interface Metadata {
    "value": string | number,
    "type": MetadataType
    "url": string | undefined
    "status": Status | undefined
}

/**
 * An associated service or topic url.
 *
 * @param name - the name of the url, e.g. grafana
 * @param value - the url, e.g. https://my.grafana.co.uk/event-stream-dashboard/my-topic
 * @param icon? - a material ui icon to render urls, if null urls are rendered as 'launch' icon
 * @type {Url}
 * */
export interface Url {
    "name": string,
    "value": string,
    "icon": string | undefined
}

/**
 * Defines a link between two nodes
 *
 * @param source - the node id of the source
 * @param target - the node id of the target
 * @type {EventStreamRegistryLink}
 * */
export interface EventStreamRegistryLink {
    source: string,
    target: string,
}

export enum MetadataType {
    "NUMBER" = "NUMBER",
    "STRING" = "STRING"
}

export enum Status {
    "BAD",
    "GOOD",
}

export enum StatusColors {
    BAD = "#FF0000",
    GOOD = "#008000",
    NO_STATUS = "#000000",
}

export enum NodeType {
    "TOPIC" = "TOPIC",
    "SERVICE" = "SERVICE"
}
