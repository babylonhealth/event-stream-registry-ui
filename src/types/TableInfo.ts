export type MetadataFields = {
    numeric: Set<string>,
    string: Set<string>,
    url: Set<string>
}

export type EventStreamRegistryMetadata = {
    topic: MetadataFields,
    service: MetadataFields
}
export type QueryParameter = {
    key: string,
    value: string
    operation: string
}
export enum View {
    "GRAPH"= "GRAPH",
    "SERVICES"= "SERVICES",
    "TOPICS"= "TOPICS",
    "HELP"= "HELP"
}

