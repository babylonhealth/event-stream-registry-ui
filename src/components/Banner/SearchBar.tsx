import * as React from "react";

import "./SearchBar.scss"


type SearchBarProps = {
    text: string,
    onTextChange: (x: React.ChangeEvent<HTMLInputElement>) => void
}

/**
 * Renders a search bar that allows users to filter nodes and change the Event Stream Registry view
 * @param {string} text - the input text
 * @param {(x) => void} onTextChange - the action triggered on a change of the input
 * @constructor
 */
const SearchBar = ({text, onTextChange}: SearchBarProps) =>
        <input type="text"
               value={text}
               placeholder="SEARCH... (or type help)"
               spellCheck={false}
               autoFocus
               aria-label={"search-bar"}
               onChange={onTextChange}/>;

export default SearchBar;
