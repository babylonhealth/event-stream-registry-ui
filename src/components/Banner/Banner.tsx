import * as React from "react";
import {View} from "../../types/TableInfo";
import {Dispatch} from "react";
import SearchBar from "./SearchBar";
import Buttons from "./Buttons";

type BannerProps = {
    queryText: string,
    onQueryTextChange: (x: React.ChangeEvent<HTMLInputElement>) => void
    view: View,
    setView: Dispatch<React.SetStateAction<View>>
}

/**
 * Renders a banner that contains the search bar and butons which allows the user to
 * to filter nodes and change the Event Stream Registry view
 * @param {string} text - the input text
 * @param {(x) => void} onQueryTextChange - the action triggered on a change of the input
 * @param {View} view - the current Event Stream Registry view
 * @param {Dispatch<React.SetStateAction<View>>} setView - callback that sets the state of view
 * @constructor
 */
const Banner = ({queryText, onQueryTextChange, view, setView}: BannerProps) =>
    <nav>
        <SearchBar onTextChange={onQueryTextChange} text={queryText}/>
        <Buttons view={view} setView={setView}/>
    </nav>;

export default Banner;
