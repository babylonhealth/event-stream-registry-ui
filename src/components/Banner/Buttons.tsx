import * as React from "react";

import {View} from "../../types/TableInfo";
import {Dispatch} from "react";

import "./Buttons.scss"


type ButtonProps = {
    icon: string,
    clicked: boolean,
    view: View,
    setView: Dispatch<React.SetStateAction<View>>
}

type ButtonsProps = {
    view: View,
    setView: Dispatch<React.SetStateAction<View>>
}

export const GRAPH_ICON = "bubble_chart";
export const TOPIC_TABLE_ICON = "fiber_manual_record";
export const SERVICE_TABLE_ICON = "crop_din";
export const HELP_ICON = "help_outline";
/**
 * Renders buttons that allows users to change the Event Stream Registry view
 * @param {View} view - the current Event Stream Registry view
 * @param {Dispatch<React.SetStateAction<View>>} setView - callback that sets the state of view
 * @constructor
 */
const Buttons = ({view, setView}: ButtonsProps) =>
    <div className={"buttons"}>
        <Button icon={GRAPH_ICON} clicked={view === View.GRAPH} view={View.GRAPH} setView={setView}/>
        <Button icon={TOPIC_TABLE_ICON} clicked={view === View.TOPICS} view={View.TOPICS} setView={setView}/>
        <Button icon={SERVICE_TABLE_ICON} clicked={view === View.SERVICES} view={View.SERVICES} setView={setView}/>
        <Button icon={HELP_ICON} clicked={view === View.HELP} view={View.HELP} setView={setView}/>
    </div>;

/**
 * Renders single button
 *
 * @param {string} icon - the material-ui icon
 * @param {boolean} clicked - determines is button is in clicked state, used in setting the className
 * @param {View} view - Event Stream Registry view associated with this button
 * @param {Dispatch<React.SetStateAction<View>>} setView - callback that sets the state of view
 * @constructor
 */
const Button = ({icon, clicked, view, setView}: ButtonProps) =>
    <span className={"material-icons" + (clicked ? " clicked" : "")}
          onClick={() => setView(view)}>{icon}
    </span>;

export default Buttons;
