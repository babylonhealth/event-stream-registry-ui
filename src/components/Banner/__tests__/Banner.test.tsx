import React from 'react'
import {cleanup, render} from '@testing-library/react'
import Banner from '../Banner'
import {View} from "../../../types/TableInfo";

afterEach(cleanup);

it('should take a snapshot', () => {
    const {asFragment} = render(<Banner queryText={""} setView={jest.fn} onQueryTextChange={jest.fn()} view={View.GRAPH}/>);
    // @ts-ignore
    expect(asFragment(<Banner queryText={""} setView={jest.fn} onQueryTextChange={jest.fn()} view={View.GRAPH}/>))
        .toMatchSnapshot()
});
