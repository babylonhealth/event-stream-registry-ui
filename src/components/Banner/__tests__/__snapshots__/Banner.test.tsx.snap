// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <nav>
    <input
      aria-label="search-bar"
      placeholder="SEARCH... (or type help)"
      spellcheck="false"
      type="text"
      value=""
    />
    <div
      class="buttons"
    >
      <span
        class="material-icons clicked"
      >
        bubble_chart
      </span>
      <span
        class="material-icons"
      >
        fiber_manual_record
      </span>
      <span
        class="material-icons"
      >
        crop_din
      </span>
      <span
        class="material-icons"
      >
        help_outline
      </span>
    </div>
  </nav>
</DocumentFragment>
`;
